product/src/main/java/com/yas/product/service/ProductService.java

```diff
@@ -285,21 +287,55 @@ private Product buildProductVariationFromVm(ProductVariationSaveVm variationVm,
     }
 
     private List<ProductOptionValue> createProductOptionValues(ProductPostVm productPostVm, Product savedMainProduct,
-                                                               Map<Long, ProductOption> optionsById) {
-        List<ProductOptionValue> optionValues = new ArrayList<>();
-        productPostVm.productOptionValues().forEach(optionValueVm -> optionValueVm.value().forEach(value -> {
-            ProductOptionValue optionValue = ProductOptionValue.builder()
-                .product(savedMainProduct)
-                .displayOrder(optionValueVm.displayOrder())
-                .displayType(optionValueVm.displayType())
-                .productOption(optionsById.get(optionValueVm.productOptionId()))
-                .value(value)
-                .build();
-            optionValues.add(optionValue);
-        }));
+            Map<Long, ProductOption> optionsById) {
+
+        List<ProductOptionValueSaveVm> productOptionValues = productPostVm.productOptionValues();
+        Map<Long, Set<String>> optionValuesMap = new HashMap<>();
+        List<ProductOptionValue> optionValues = createOptionValuesList(
+                productOptionValues,
+                optionValuesMap,
+                savedMainProduct,
+                optionsById);
         return productOptionValueRepository.saveAll(optionValues);
     }
 
+    private List<ProductOptionValue> createOptionValuesList(
+            List<ProductOptionValueSaveVm> productOptionValues,
+            Map<Long, Set<String>> optionValuesMap,
+            Product product,
+            Map<Long, ProductOption> optionsById
+    ) {
+        List<ProductOptionValue> optionValues = new ArrayList<>();
+
+        for (ProductOptionValueSaveVm productOptionValueSaveVm : productOptionValues) {
+            Long productOptionId = productOptionValueSaveVm.productOptionId();
+            List<String> values = productOptionValueSaveVm.value();
+
+            Set<String> set;
+            if (optionValuesMap.containsKey(productOptionId)) {
+                set = optionValuesMap.get(productOptionId);
+            } else {
+                set = new HashSet<>();
+                optionValuesMap.put(productOptionId, set);
+            }
```

Comment:
```
Set<String> set;
            if (optionValuesMap.containsKey(productOptionId)) {
                set = optionValuesMap.get(productOptionId);
            } else {
                set = new HashSet<>();
                optionValuesMap.put(productOptionId, set);
            }
```
I think this code is quite confusing and duplicated. Why don't you filter the values of `List<ProductOptionValueSaveVm>` in `updateProductOptionValues  `before passing it to this method? So the `Map<Long, Set<String>> optionValuesMap` is unnecessary here.

Reply:
if I filter the values of `List<ProductOptionValueSaveVm>` before, I will still need to loop though the `List` inside the method, thus loop `for (ProductOptionValueSaveVm productOptionValueSaveVm : productOptionValues)` would be used twice (one for filter, one for insert). This code is actually the shorten form of the two loops. Besides, I still need to use set to filter `ProductOptionValueSaveVm` for unique option values, right?

Reply:
I implemented a sample code for filtering List<ProductOptionValueSaveVm>. Please help to check and test to see if you can use it.

```
List<ProductOptionValueSaveVm> productOptionValues = productPutVm.productOptionValues();
        List<ProductOptionValue> existingProductOptionValues = productOptionValueRepository.findAllByProduct(product);
        Map<Long, Set<String>> existingOptionValuesMap = existingProductOptionValues.stream().collect(
                Collectors.groupingBy(ProductOptionValue::getId, Collectors.mapping(ProductOptionValue::getValue, Collectors.toSet())));

        for (ProductOptionValueSaveVm productOptionValueSaveVm : productOptionValues) {
            Set<String> existingValues = existingOptionValuesMap.get(productOptionValueSaveVm.productOptionId());
            if (CollectionUtils.isNotEmpty(existingValues)) {
                existingValues.addAll(productOptionValueSaveVm.value());
                productOptionValueSaveVm.value().clear();
                productOptionValueSaveVm.value().addAll(existingValues);
            }
        }
```

Reply:
I think this won't work, because for update I actually need unique, not yet saved option values, 
`productOptionValueSaveVm.value().addAll(existingValues);` would contain all unique saved and not saved values, so I would need to have another `for` loop to filter saved values out.
Also, instead of 
```
[
  {
    productOptionId: 1, 
    value: ["Red", "Green"]
  },
  {
    productOptionId: 2,
    value: ["8GB", "16GB"]
  }
]
```
the storefront is currently sent:
```
[
  {
    productOptionId: 1, 
    value: ["Red"]
  },
  {
    productOptionId: 1, 
    value: ["Green"]
  },
  {
    productOptionId: 2,
    value: ["8GB"]
  }
  {
    productOptionId: 2,
    value: ["16GB"]
  }
]
```
_displayType and displayOrder omitted_
That why I retrieve the set and check uniqueness for each value
So, I don't think the code snippet will work

Reply:
Can you check my sample code more carefully? Of course, I haven't tested it but I'm sure I understand your problem. I need you to help test my suggestion.

```
Map<Long, Set<String>> existingOptionValuesMap = existingProductOptionValues.stream().collect(
                Collectors.groupingBy(ProductOptionValue::getId, Collectors.mapping(ProductOptionValue::getValue, Collectors.toSet())));
```
This block of code above groups values ​​of the same `productOptionId` into a `Set`. I think the result will be as your expectation.

> [
>   {
>     productOptionId: 1, 
>     value: ["Red", "Green"]
>   },
>   {
>     productOptionId: 2,
>     value: ["8GB", "16GB"]
>   }
> ]